### DISCUSSION TITLE
Security improvement of SQL query

### DISCUSSION THREAD

**Discussion by ColonelMoutarde** [FreshRSS/FreshRSS#5026] [General]
This code is from https://insight.symfony.com/projects/c0384218-7393-45e4-83f8-c95491232740/analyses/22

Your project database queries must use safety features such as bindings 24

in [app/Models/CategoryDAO.php](https://github.com/ColonelMoutarde/FreshRSS/blob/db05ecdf677429b353f905ece2eb5fe5e84cf845/app/Models/CategoryDAO.php#L315), line 315
```php
    	}
    	/** @return array<FreshRSS_Category> */
    	public function listCategoriesOrderUpdate(int $defaultCacheDuration = 86400, int $limit = 0): array {
    		$sql = 'SELECT * FROM `_category` WHERE kind = :kind AND `lastUpdate` < :lu ORDER BY `lastUpdate`'
    			. ($limit < 1 ? '' : ' LIMIT ' . $limit);
```

If provided by the user, the value of `$limi`t may allow an SQL injection attack. Avoid concatenating parameters to SQL query strings, and use parameter binding instead.

The project could be improved in the same way in other requests
[https://insight.symfony.com/what-we-analyse/sample](https://insight.symfony.com/what-we-analyse/sample)

**Comment by Frenzie** [comment]
It says $limit must be an integer so I don't think it may allow much of anything that would be prevented by using bindParam. But yes, agreed that it'd be better practice to avoid potential issues like that from accidentally sneaking in.

**Reply by ColonelMoutarde** [reply]
Like that https://github.com/ColonelMoutarde/FreshRSS/pull/1/files

**Comment by Alkarex** [comment]
We used to have an `intval()` there, but it was removed by @ColonelMoutarde in https://github.com/FreshRSS/FreshRSS/pull/4561
The check is now done by the function type declaration to int as @Frenzie is saying, so no security problem there.

**Reply by Frenzie** [reply]
I think the argument we can derive from the OP is that it requires less reasoning if you simply do them all with `bindValue()`, which would also make it harder to accidentally introduce an issue in the future. Like if someone copy pastes it and duplicates the concatenation, but changes it to a `string $bleep`. Or just so you don't have to double check whether it's safe or not when skimming. I'm not entirely sure where I stand on that argument, but in any case I think it's ultimately about clarity at a glance and not whether there's an active issue.

The argument works better if you imagine the code in the OP as separated from the function declaration by a few dozen lines though. :-)

**Reply by Alkarex** [reply]
Sure. My points were that:
* No security problem there (verified int, and not a user input anyway)
* It was clear with `intval()` before https://github.com/FreshRSS/FreshRSS/pull/4561

**Reply by ColonelMoutarde** [reply]
I understand your answer, but good development practices forbid concatenation in sql queries. 
The correction is trivial.
 Let's apply guidelines recognized in the profession.

**Reply by Alkarex** [reply]
Again, I am not arguing about a bind, just the wrong title and analysis, coupled by the fact that this should have been done by https://github.com/FreshRSS/FreshRSS/pull/4561
PR welcome, but https://github.com/ColonelMoutarde/FreshRSS/pull/1 is not good due to duplicated code. Even better if searching in the codebase for a similar pattern.
